Quickstart flutter mobile (#2514)
FusionAuth/fusionauth-site: astro/src/content/quickstarts/quickstart-flutter-native.mdx

<Aside type="note">
If you ever want to reset the FusionAuth system, delete the volumes created by `docker-compose` by executing `docker-compose down -v`, then rerun `docker-compose up -d`.
</Aside>

FusionAuth will be configured with these settings:

* Your client Id is `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is `super-secret-secret-that-should-be-regenerated-for-production`.
* Your example username is `richard@example.com` and the password is `password`.
* Your admin username is `admin@example.com` and the password is `password`.
* The base URL of FusionAuth is `http://localhost:9011/`.

You can log in to the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want to, but with Docker and Kickstart, everything will already be configured correctly.

<Aside type="caution">
  The `.env` and `kickstart.json` files contain passwords. In a real application, always add these files to your `.gitignore` file and never commit secrets to version control.
</Aside>

> ### Set Up A Public URL For FusionAuth

Reviewer: Could we add some verbiage here to include this as an easy guide?
https://fusionauth.io/docs/v1/tech/developer-guide/exposing-instance

Then include this script
```shell
ngrok http --request-header-add 'X-Forwarded-Port:443' 9011
```

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-flutter-native/main/complete-application/android/app/build.gradle"
 lang="text" tags="appAuthRedirectScheme" />

#### iOS Setup

Now edit the `ios/Runner/Info.plist` file in the iOS app to specify the custom scheme. Find the section that looks similar to the following and add the FusionAuth URL `com.fusionauth.flutterdemo`.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-flutter-native/main/complete-application/ios/Runner/Info.plist"
 lang="xml" />

Make sure the `ios/Runner/Info.plist` file has `CFBundleURLTypes` and `CFBundleURLSchemes` keys as in the code above.

### Dive Into The Code

Open the `main.dart` file in the `lib` directory of your project and paste the contents below into it.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-flutter-native/main/complete-application/lib/main.dart"
 lang="dart" />

<Aside type="tip">
Putting all your logic in one file makes sense for a tutorial but for a larger application, you'll probably want to split it up.
</Aside>

> At the top of the file, change the `FUSIONAUTH_DOMAIN` constant to the public URL for your FusionAuth instance (the URL you used when configuring it).

Reviewer: Include link back to `/#set-up-a-public-url-for-fusionauth` for clarity.

To make sure your Flutter development environment is working correctly, run the following command in your terminal window.

```shell
flutter doctor
```

If everything is configured properly, you will see something like the following result in your terminal window:

```shell
Doctor summary (to see all details, run flutter doctor -v):
[✓] Flutter (Channel stable, 3.13.12)
[✓] Android toolchain - develop for Android devices (Android SDK version 34.0.0)
[✓] Xcode - develop for iOS and macOS (Xcode 14.3.1)
[✓] Chrome - develop for the web
[✓] Android Studio (version 2022.3)
[✓] IntelliJ IDEA Community Edition (version 2022.2)
[✓] Connected device (2 available)
[✓] Network resources

• No issues found!
```

## General Architecture

While this sample application doesn't have login functionality without FusionAuth, a more typical integration will replace an existing login system with FusionAuth.

In that case, the system might look like this before FusionAuth is introduced.

<LoginBefore alt={"Request flow during login before FusionAuth"}/>

The login flow will look like this after FusionAuth is introduced.

<LoginAfter alt={"Request flow during login after FusionAuth"}/>

> In general, you would introduce FusionAuth to normalize and consolidate user data, making it consistent and up-to-date and offloading your login security and functionality to FusionAuth.

Reviewer 2: ```suggestion
In general, you would introduce FusionAuth to normalize and consolidate user data, ensuring consistency and reliable, while offloading your login security and functionality to FusionAuth.
```

## Getting Started

> Start with getting FusionAuth up and running and creating a Flutter application.

Reviewer 2: ```suggestion
Start by getting FusionAuth up and running and creating a Flutter application.
```

- [Linux](https://linuxize.com/post/how-to-find-ip-address-linux/#find-your-public-ip-address)
- [Mac](https://support.apple.com/guide/mac-help/find-your-computers-name-and-network-address-mchlp1177/mac)
- [Windows](https://support.microsoft.com/en-us/windows/find-your-ip-address-in-windows-f21a9bbc-c582-55cd-35e0-73431160a1b9)

### Create Your Flutter Application

In the directory where the project should live, run the following command to create and set up the new Flutter app.

```shell
flutter create --org com.fusionauth flutterdemo --platforms=ios,android
```

After the installation process completes, you will see that the `flutterdemo` directory contains all the Flutter starter app configuration. Open the project directory with the text editor of your choice.

You can run your new project on the actual device or an emulator to confirm everything is working before you customize any code. Do so by running the following command in the project directory.

```shell
cd flutterdemo
flutter run
```

> From the list of emulators or devices that is returned, choose the one you want the code to run in. Stay in this directory for the rest of this tutorial.

Reviewer 2: ```suggestion
From the list of emulators or devices that can be used, choose the one that you want to run the code on.
Stay in this directory for the rest of this tutorial.
```
